tcl/file.tcl

```diff
@@ -55,41 +55,51 @@ proc qc::file_upload {name chunk chunks file} {
 
     # Move the AOLserver generated tmp file to one we will keep
     file rename [ns_getformfile file] $tmp_file
-    if { [nsv_exists uploads $user_id] } {
-	set dict [nsv_get uploads $user_id]
-    } else {
-	set dict {}
-    }
-    dict set dict $id $chunk $tmp_file
-    dict set dict $id chunks $chunks
-    dict set dict $id filename $file
-
-    set complete true
-    foreach chunk [.. 0 $chunks-1] {
-	if { ![dict exists $dict $id $chunk] } {
-	    set complete false
-	    break
-	} else {
-	    lappend files /tmp/$id.$chunk
-	}
+
+    if { ![nsv_exists locks $user_id] } {
+        nsv_set locks $user_id [ns_mutex create]
```

Comment:
We create a lock for each user and use an NSV array to track them.

It perhaps would be better to go a layer further and have a lock per file per user but as far as I'm aware there's no guarantee that the file ID generated by the caller is unique. It would also required storing a dict as the value, checking lock existence in the dict, and updating the NSV with a new dict. Naviserver 4.99.21+ supports [nsv_dict](https://naviserver.sourceforge.io/n/naviserver/files/nsv.html#8) which would be more suited for this purpose.

A lock per user has been sufficient in my testing so far.